buffer/src/main/java/io/netty/buffer/AdaptivePoolingAllocator.java

```diff
@@ -314,120 +305,306 @@ void reallocate(int size, int maxCapacity, AdaptiveByteBuf into) {
-    // we might end up with leaks. While these leaks are usually harmless in reality it would still at least be
-    // very confusing for users.
-    @Override
-    protected void finalize() throws Throwable {
-        try {
-            super.finalize();
-        } finally {
-            free();
+        private void freeCentralQueue() {
+            for (;;) {
+                Chunk chunk = chunkReuseQueue.poll();
+                if (chunk == null) {
+                    break;
+                }
+                chunk.release();
+            }
         }
     }
 
-    private void free() {
-        freed = true;
-        long stamp = magazineExpandLock.writeLock();
-        try {
-            Magazine[] mags = magazines;
-            for (Magazine magazine : mags) {
-                magazine.free();
-            }
-        } finally {
-            magazineExpandLock.unlockWrite(stamp);
+    private interface ChunkControllerFactory {
+        ChunkController create(MagazineGroup group);
+    }
+
+    private interface ChunkController {
+        /**
+         * Compute the "fast max capacity" value for the buffer.
+         */
+        int computeBufferCapacity(int requestedSize, int maxCapacity, boolean isReallocation);
+
+        /**
+         * Initialize the given chunk factory with shared statistics state (if any) from this factory.
+         */
+        void initializeSharedStateIn(ChunkController chunkController);
+
+        /**
+         * Allocate a new {@link Chunk} for the given {@link Magazine}.
+         */
+        Chunk newChunkAllocation(int promptingSize, Magazine magazine);
+    }
+
+    private interface ChunkReleasePredicate {
+        boolean shouldReleaseChunk(int chunkSize);
+    }
+
+    private static final class SizeClassChunkControllerFactory implements ChunkControllerFactory {
+        private final int segmentSize;
+
+        private SizeClassChunkControllerFactory(int segmentSize) {
+            this.segmentSize = ObjectUtil.checkPositive(segmentSize, "segmentSize");
+        }
+
+        @Override
+        public ChunkController create(MagazineGroup group) {
+            return new SizeClassChunkController(group, segmentSize);
         }
-        freeCentralQueue();
     }
 
-    private void freeCentralQueue() {
-        for (;;) {
-            Chunk chunk = centralQueue.poll();
-            if (chunk == null) {
-                break;
-            }
-            chunk.release();
+    private static final class SizeClassChunkController implements ChunkController {
+        private final MagazineGroup group;
+        private final int segmentSize;
+        private final int chunkSize;
+
+        private SizeClassChunkController(MagazineGroup group, int segmentSize) {
+            this.group = group;
+            this.segmentSize = segmentSize;
+            chunkSize = Math.max(MIN_CHUNK_SIZE, segmentSize * 32);
```

Comment:
Can you add a comment why we use `32` here or make this magic number a static field with a comment ?